keras_nlp/models/gpt_neo_x/rotary_embedding.py

```diff
@@ -11,49 +11,53 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-import tensorflow as tf
-
 from keras_nlp.backend import keras
+from keras_nlp.backend import ops
 
 
 class RotaryEmbedding(keras.layers.Layer):
     def __init__(self, max_wavelength=10000, **kwargs):
         super().__init__(**kwargs)
         self.max_wavelength = max_wavelength
 
-    def _apply_rotary_pos_emb(self, tensor, cos_emb, sin_emb):
-        cos_emb = cos_emb[:, : tf.shape(tensor)[1], :, :]
-        sin_emb = sin_emb[:, : tf.shape(tensor)[1], :, :]
+    def build(self, rotary_dim):
+        self.rotary_dim = rotary_dim
+        self.built = True
 
-        x1, x2 = tf.split(tensor, 2, axis=-1)
-        half_rot_tensor = tf.concat((-x2, x1), axis=-1)
+    def __call__(self, query, key):
+        if not self.built:
+            rotary_dim = query.shape[-1]
+            self.build(rotary_dim)
+        return super().__call__(query, key)
 
-        return (tensor * cos_emb) + (half_rot_tensor * sin_emb)
+    def call(self, query, key):
+        cos_emb, sin_emb = self._compute_cos_sin_embedding(key, seq_dim=1)
+        query = self._apply_rotary_pos_emb(query, cos_emb, sin_emb)
+        key = self._apply_rotary_pos_emb(key, cos_emb, sin_emb)
 
-    def _compute_cos_sin_embedding(self, x, rotary_dim, seq_dim=1):
-        seq_len = tf.shape(x)[seq_dim]
-        rotary_dim = tf.cast(rotary_dim, "float32")
+        return query, key
 
-        range = tf.range(start=0, limit=rotary_dim, delta=2, dtype="float32")
-        inverse_freq = 1.0 / (self.max_wavelength ** (range / rotary_dim))
+    def _apply_rotary_pos_emb(self, tensor, cos_emb, sin_emb):
+        cos_emb = cos_emb[:, : ops.shape(tensor)[1], :, :]
+        sin_emb = sin_emb[:, : ops.shape(tensor)[1], :, :]
 
-        tensor = tf.range(seq_len, dtype=inverse_freq.dtype)
-        freqs = tf.einsum("i, j -> ij", tensor, inverse_freq)
-        embedding = tf.concat((freqs, freqs), axis=-1)[None, :, None, :]
+        x1, x2 = ops.split(tensor, 2, axis=-1)
+        half_rot_tensor = ops.concatenate((-x2, x1), axis=-1)
 
-        return tf.cos(embedding), tf.sin(embedding)
+        return (tensor * cos_emb) + (half_rot_tensor * sin_emb)
 
-    def call(self, query, key):
-        rotary_dim = tf.shape(query)[-1]
+    def _compute_cos_sin_embedding(self, x, seq_dim=1):
+        seq_len = ops.shape(x)[seq_dim]
+        rotary_dim = ops.cast(self.rotary_dim, "float32")
```

Comment:
probably better to use `self.compute_dtype` instead of `float32`, that would cover the mixed precision case.

Reply:
cool !